File: objectiveai-sdk-js/src/laboratories/executions/http.ts
import { ObjectiveAI, type RequestOptions } from "../../client";
import { Stream } from "../../stream";
import type { LaboratoriesExecutionsRequestLaboratoryExecutionCreateParams } from "./request/laboratoryExecutionCreateParams";
import type { LaboratoriesExecutionsResponseUnaryLaboratoryExecution } from "./response/unary/laboratoryExecution";
import type { LaboratoriesExecutionsResponseStreamingLaboratoryExecutionChunk } from "./response/streaming/laboratoryExecutionChunk";

export function laboratoriesExecutionsCreateLaboratoryExecution(
  client: ObjectiveAI,
  body: LaboratoriesExecutionsRequestLaboratoryExecutionCreateParams & { stream: true },
  options?: RequestOptions,
): Promise<Stream<LaboratoriesExecutionsResponseStreamingLaboratoryExecutionChunk>>;
export function laboratoriesExecutionsCreateLaboratoryExecution(
  client: ObjectiveAI,
  body: LaboratoriesExecutionsRequestLaboratoryExecutionCreateParams & { stream?: false | null },
  options?: RequestOptions,
): Promise<LaboratoriesExecutionsResponseUnaryLaboratoryExecution>;
export function laboratoriesExecutionsCreateLaboratoryExecution(
  client: ObjectiveAI,
  body: LaboratoriesExecutionsRequestLaboratoryExecutionCreateParams,
  options?: RequestOptions,
): Promise<
  | Stream<LaboratoriesExecutionsResponseStreamingLaboratoryExecutionChunk>
  | LaboratoriesExecutionsResponseUnaryLaboratoryExecution
> {
  if (body.stream) {
    return client.post_streaming<LaboratoriesExecutionsResponseStreamingLaboratoryExecutionChunk>(
      "laboratories/executions",
      body,
      options,
    );
  }
  return client.post_unary<LaboratoriesExecutionsResponseUnaryLaboratoryExecution>(
    "laboratories/executions",
    body,
    options,
  );
}

File: objectiveai-sdk-py/objectiveai_sdk/laboratories/executions/http.py
"""HTTP functions for laboratory executions."""

from __future__ import annotations

from typing import TYPE_CHECKING, Union

if TYPE_CHECKING:
    from objectiveai.client import ObjectiveAI
    from objectiveai.laboratories.executions.request import LaboratoryExecutionCreateParams
    from objectiveai.laboratories.executions.response.streaming import LaboratoryExecutionChunk
    from objectiveai.laboratories.executions.response.unary import LaboratoryExecution
    from objectiveai.stream import Stream


async def create_laboratory_execution(
    client: ObjectiveAI,
    params: LaboratoryExecutionCreateParams,
) -> Union[LaboratoryExecution, Stream[LaboratoryExecutionChunk]]:
    """Execute a laboratory.

    If ``params.stream`` is true, returns a streaming response.
    """
    if getattr(params, "stream", None):
        return await client.post_streaming("laboratories/executions", params)
    return await client.post_unary("laboratories/executions", params)
